tests/unit/packet/PublishTest.php

```diff
@@ -15,134 +16,135 @@ public function testPublishStandard()
         $this->assertEquals(3, Publish::getControlPacketType());
     }
 
+    public function testExceptionIsThrownForUnexpectedPacketType()
+    {
+        $input =
+            chr(0b00100000) .
+            chr(2) .
+            chr(0) .
+            chr(0);
+
+        $this->setExpectedException(
+            'RuntimeException',
+            'raw input is not valid for this control packet'
+        );
+
+        Publish::parse(new Version4(), $input);
+    }
+
     public function testPublishStandardWithQos2()
     {
-        $version = new \oliverlorenz\reactphpmqtt\protocol\Version4();
-        $packet = new Publish($version);
+        $packet = new Publish(new Version4());
         $packet->setQos(2);
 
-        $this->assertEquals(
-            MessageHelper::getReadableByRawString(
-                chr(52) .
-                chr(2) .
-                chr(0) .
-                chr(0) .
-                chr(49)
-            ),
-            MessageHelper::getReadableByRawString($packet->get() . 1)
-        );
+        $expected =
```

Comment:
can you add the documentation link for QOS here? It was very hard to understand what is going on here. A link to the documentation will give a starting point. Please add this links also to the other functions because its not easy to understand what the bytes stand for.

Thank you! :+1: